tensorboard/components/tf_data_selector/tf-data-select-row.ts

```diff
@@ -14,64 +14,189 @@ limitations under the License.
+      value: false,
+    },
 
-    selectedRuns: {
+    _runs: {
       type: Array,
-      notify: true,
       value: () => [],
     },
 
-    _runSelectionState: {
-      type: Object,
-      observer: '_storeRunSelectionState',
-      value: () => tf_storage.getObject('runSelectionState') || {},
+    _tags: {
+      type: Array,
+      value: () => [],
+    },
+
+    selection: {
+      type: Array,
+      notify: true,
+      computed: '_computeSelection(_selectedRuns, _selectedTags)',
     },
 
-    _runRegexInput: {
-      type: String,
-      value: tf_storage.getStringInitializer('regexInput', {
-            defaultValue: '',
-            polymerProperty: '_runRegexInput',
-          }),
-      observer: '_storeRunRegexInput',
+    _runSelectionStateString: {type: String, value: ''},
+
+    _selectedRuns: {
+      type: Array,
+      value: () => [],
+    },
+
+    _tagSelectionStateString: {type: String, value: ''},
+
+    _selectedTags: {
+      type: Array,
+      value: () => [],
     },
   },
 
-  get _runsStore() {
-    return tf_backend.runsStore;
+  observers: [
+    '_persistSelectedRuns(_selectedRuns)',
+    '_persistSelectedTags(_selectedTags)',
+  ],
+
+  _getPersistenceKey(type: Type): string {
+    const number = this.persistenceNumber || 0;
+    switch (type) {
+      case Type.RUN:
+        return `g${number}r`;
+      case Type.TAG:
+        return `g${number}t`;
+    }
+  },
+
+  ready() {
+    if (this.persistenceNumber == null) return;
+
+    const runInitializer = tf_storage.getStringInitializer(
+        this._getPersistenceKey(Type.RUN),
+        {defaultValue: '', polymerProperty: '_runSelectionStateString'});
+    runInitializer.call(this);
+
+    const tagInitializer = tf_storage.getStringInitializer(
+        this._getPersistenceKey(Type.TAG),
+        {defaultValue: '', polymerProperty: '_tagSelectionStateString'});
+    tagInitializer.call(this);
   },
 
   attached() {
-    this._updateRunKey = this._runsStore.addListener(() => this._updateRuns());
-    this._updateRuns();
+    this._fetchRunsAndTags().then(() => this._isDataReady = true);
   },
 
   detached() {
-    this._runsStore.removeListenerByKey(this._updateRunKey);
+    this._isDataReady = false;
   },
 
-  _updateRuns() {
-    this.set('_runs', this._runsStore.getRuns());
+  _fetchRunsAndTags() {
+    const requestManager = new tf_backend.RequestManager();
+    if (this.noExperiment) {
+      const fetchRuns = requestManager.request(tf_backend.getRouter().runs());
+      return Promise.all([fetchRuns]).then(([runs]) => {
+        this.set('_runs', Array.from(new Set(runs)).map(runName => ({
+          id: runName,
+          name: runName,
+          startedTime: null,
+        })));
+      });
+    } else if (this.experiment.id) {
+      const url = tf_backend.getRouter().runsForExperiment(this.experiment.id);
+      return requestManager.request(url).then(runs => {
+        this.set('_runs', runs);
+        // Flatten the tags.
```

Comment:
This isn't quite how tag IDs work.  A tag ID in the SQLite schema is unique across all tags/runs/etc - it corresponds to a particular instance of a "tag name" for a specific run.  I think when you're flattening this you actually want to collapse the tag names themselves into a unique set rather than deduping by ID (which would be a no-op).